File: Metodos/Probabilidad/Metropolis/markov1D.py
import numpy as np
import matplotlib.pyplot as plt

N = 10**6 # Numero de pasos
x0 = 1 # Origen
delta = 0.1 # El paso va a tener una longitud en [-delta,delta]
sigma = 1

x = np.array([x0],dtype='float')

def q(x1,x2): # Probabilidad de pasar a x1 dado que estoy en x2
    return 1 # Vamos a asumir que hay la misma probabilidad de saltar a la derecha o a la izquierda

def p(x): # Probabilidad de estar en la posicion x
    return (1/(np.sqrt(2*np.pi)*sigma))*np.exp(-x**2/(2*sigma**2))

def U(): # Un numero aleatorio que va a decidir si se da el salto o no
    return np.random.rand()

for i in range(N-2):
    xnew = x[i] + (np.random.rand() - 0.5)*2*delta # Posible posicion nueva
    alpha = min(1,(p(xnew)/p(x[i]))*(q(xnew,x[i])/q(x[i],xnew)))
    u = U()
    if (u < alpha):
        x = np.append(x,xnew)
    else:
        x = np.append(x,x[i])

plt.hist(x,bins=50)
plt.show()

File: Metodos/Probabilidad/Metropolis/markov1D.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define Pi 3.141592653589793
#define Sigma 1
#define delta 0.2
#define N (int)pow(10,6)
#define x0 1

double p(double x);

int main(void){
  int i;
  double x = x0;
  double xnew;
  double alpha;
  double u;
  srand48(1);
  printf("%f\n",x);
  for (i=0;i<N;i++){
    xnew = x + drand48()*2*delta - delta;
    if (1 < p(xnew)/p(x)){
      alpha = 1;
    }
    else{
      alpha = p(xnew)/p(x);
    }
    u = drand48();
    if (u < alpha){
      x = xnew;
    }
    printf("%f\n",x);
  }
  return 0;
}

double p(double x){
  return((1.0/(Sigma*pow(2*Pi,0.5)))*exp(-pow(x,2)/(2*pow(Sigma,2))));
}
